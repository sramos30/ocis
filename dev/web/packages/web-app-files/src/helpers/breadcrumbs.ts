import { bus } from 'web-pkg/src/instance'

export interface BreadcrumbItem {
  text: string
  to?: string
  allowContextActions?: boolean
  onClick?: () => void
}

export const breadcrumbsFromPath = (
  currentPath: string,
  resourcePath: string
): BreadcrumbItem[] => {
  const pathSplit = (p = '') => p.split('/').filter(Boolean)
  const current = pathSplit(currentPath)
  const resource = pathSplit(resourcePath)

  return resource.map(
    (text, i) =>
      ({
        allowContextActions: true,
        text,
        to: '/' + [...current].splice(0, current.length - resource.length + i + 1).join('/')
      } as BreadcrumbItem)
  )
}

export const concatBreadcrumbs = (...items: BreadcrumbItem[]): BreadcrumbItem[] => {
  const last = items.pop()

  return [
    ...items,
    {
      allowContextActions: last.allowContextActions,
      text: last.text,
      onClick: () => bus.publish('app.files.list.load')
    }
  ]
}
