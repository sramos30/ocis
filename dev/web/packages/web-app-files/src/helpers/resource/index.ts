export * from './asset'
export * from './filter'
export * from './privatePreviewBlob'
export * from './publicPreviewUrl'
export * from './resource'
export * from './sameResource'
