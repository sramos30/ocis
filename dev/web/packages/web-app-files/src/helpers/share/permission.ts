// dummy to trick gettext string extraction into recognizing strings
const $gettext = (str) => {
  return str
}

export class SharePermission {
  private readonly _key: string
  private readonly _bit: number
  private readonly _label: string

  constructor(key: string, bit: number, label: string) {
    this._key = key
    this._bit = bit
    this._label = label
  }

  get key(): string {
    return this._key
  }

  get bit(): number {
    return this._bit
  }

  get label(): string {
    return $gettext(this._label)
  }

  public enabled(bitmask: number): boolean {
    return (bitmask & this.bit) > 0
  }
}

export abstract class SharePermissions {
  static readonly read = new SharePermission('read', 1, $gettext('Read'))
  static readonly update = new SharePermission('update', 2, $gettext('Update'))
  static readonly create = new SharePermission('create', 4, $gettext('Create'))
  static readonly delete = new SharePermission('delete', 8, $gettext('Delete'))
  static readonly share = new SharePermission('share', 16, $gettext('Share'))

  static permissionsToBitmask(permissions: SharePermission[]): number {
    return (permissions || []).reduce((b: number, p: SharePermission) => b | p.bit, 0)
  }

  static bitmaskToPermissions(bitmask: number): SharePermission[] {
    return [this.read, this.update, this.create, this.delete, this.share].reduce(
      (a: SharePermission[], p: SharePermission) => {
        if (p.enabled(bitmask)) {
          a.push(p)
        }
        return a
      },
      []
    )
  }
}
