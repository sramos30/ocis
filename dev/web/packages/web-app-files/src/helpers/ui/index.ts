export * as fileList from './filesList'
export * as resourceTable from './resourceTable'
