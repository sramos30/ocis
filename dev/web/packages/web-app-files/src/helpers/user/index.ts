export * from './avatarUrl'
export * from './types'
