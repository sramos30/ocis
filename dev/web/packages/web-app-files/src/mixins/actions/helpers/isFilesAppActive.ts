import { activeApp } from 'web-pkg/src/composables'

const isFilesAppActive = (activeApp: string): boolean => {
  // FIXME: we should use this constant but it somehow breaks the unit tests
  // return activeApp === FilesApp.appInfo.id
  return activeApp === 'files'
}

export default {
  computed: {
    $_isFilesAppActive(): boolean {
      return isFilesAppActive(activeApp(this.$route))
    }
  }
}
