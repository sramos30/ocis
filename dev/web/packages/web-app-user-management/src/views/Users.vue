<template>
  <div>
    <create-user-modal
      v-if="createUserModalOpen"
      :existing-users="users"
      @cancel="toggleCreateUserModal"
      @confirm="createUser"
    />
    <delete-user-modal
      v-if="deleteUserModalOpen"
      :users="selectedUsers"
      @cancel="toggleDeleteUserModal"
      @confirm="deleteUsers"
    />
    <main class="oc-flex oc-flex-column oc-height-1-1 oc-p-m">
      <app-loading-spinner v-if="loadResourcesTask.isRunning" />
      <template v-else>
        <div class="oc-app-bar">
          <div class="oc-flex oc-flex-between">
            <oc-breadcrumb class="oc-flex oc-flex-middle" :items="breadcrumbs" />
          </div>
          <div class="oc-flex-1 oc-flex oc-flex-start oc-mt-m">
            <div v-if="selectedUsers.length" class="oc-flex oc-flex-middle">
              <span v-text="selectedUsersText" />
              <oc-button
                id="files-clear-selection"
                v-oc-tooltip="$gettext('Clear selection')"
                :aria-label="$gettext('Clear selection')"
                class="oc-ml-m"
                appearance="outline"
                @click="unselectAllUsers"
              >
                <oc-icon name="close" />
              </oc-button>
              <oc-button appearance="outline" class="oc-ml-m" @click="toggleDeleteUserModal">
                <oc-icon name="delete-bin" />
                <translate>Delete</translate>
              </oc-button>
            </div>
            <div v-else>
              <oc-button variation="primary" appearance="filled" @click="toggleCreateUserModal">
                <oc-icon name="add" />
                <translate>New user</translate>
              </oc-button>
            </div>
          </div>
        </div>
        <no-content-message
          v-if="!users.length"
          id="user-management-users-empty"
          class="files-empty"
          icon="user"
        >
          <template #message>
            <span v-translate>No users in here</span>
          </template>
        </no-content-message>
        <div v-else>
          <UsersList
            :users="users"
            :roles="roles"
            :user-assignments="userAssignments"
            :selected-users="selectedUsers"
            class="oc-mt-m"
            @toggleSelectUser="toggleSelectUser"
            @toggleSelectAllUsers="toggleSelectAllUsers"
          />
        </div>
      </template>
    </main>
  </div>
</template>

<script>
import UsersList from '../components/Users/UsersList.vue'
import CreateUserModal from '../components/Users/CreateUserModal.vue'
import DeleteUserModal from '../components/Users/DeleteUserModal.vue'
import AppLoadingSpinner from 'web-pkg/src/components/AppLoadingSpinner.vue'
import NoContentMessage from 'web-pkg/src/components/NoContentMessage.vue'
import { useStore } from 'web-pkg/src/composables'
import { ref } from '@vue/composition-api'
import { clientService } from 'web-pkg/src/services'
import { useTask } from 'vue-concurrency'
import { bus } from 'web-pkg/src/instance'
import { mapActions } from 'vuex'
import axios from 'axios'

export default {
  components: { UsersList, AppLoadingSpinner, NoContentMessage, CreateUserModal, DeleteUserModal },
  setup() {
    const store = useStore()
    const users = ref([])
    const roles = ref([])
    const userAssignments = ref([])
    const graphClient = clientService.graphAuthenticated(
      store.getters.configuration.server,
      store.getters.getToken
    )

    const loadRolesTask = useTask(function* (signal, ref) {
      const rolesResponse = yield axios.post(
        '/api/v0/settings/roles-list',
        {},
        {
          headers: {
            authorization: store.getters.getToken
          }
        }
      )
      roles.value = rolesResponse.data.bundles
    })

    const loadUserAssignmentTask = useTask(function* (signal, ref) {
      const userAssignmentResponse = yield axios.post(
        '/api/v0/settings/assignments-list',
        {
          account_uuid: ref.user?.id
        },
        {
          headers: {
            authorization: store.getters.getToken
          }
        }
      )
      userAssignments.value.push(userAssignmentResponse.data?.assignments)
    })

    const loadResourcesTask = useTask(function* (signal, ref) {
      const usersResponse = yield graphClient.users.listUsers('displayName')
      users.value = usersResponse.data.value || []

      yield loadRolesTask.perform()

      for (const user of users.value) {
        yield loadUserAssignmentTask.perform({ user })
      }
    })

    return {
      users,
      roles,
      userAssignments,
      loadResourcesTask,
      graphClient
    }
  },
  data: function () {
    return {
      selectedUsers: [],
      createUserModalOpen: false,
      deleteUserModalOpen: false
    }
  },
  computed: {
    selectedUsersText() {
      const translated = this.$gettext('%{ userCount } selected')

      return this.$gettextInterpolate(translated, { userCount: this.selectedUsers.length })
    },
    breadcrumbs() {
      return [
        { text: this.$gettext('User management'), to: { path: '/user-management' } },
        {
          text: this.$gettext('Users'),
          onClick: () => bus.publish('app.user-management.list.load')
        }
      ]
    },

    allUsersSelected() {
      return this.users.length === this.selectedUsers.length
    }
  },

  async mounted() {
    await this.loadResourcesTask.perform(this)

    const loadResourcesEventToken = bus.subscribe('app.user-management.list.load', () => {
      this.loadResourcesTask.perform(this)
    })

    this.$on('beforeDestroy', () => {
      bus.unsubscribe('app.user-management.list.load', loadResourcesEventToken)
    })
  },

  methods: {
    ...mapActions(['showMessage']),
    toggleSelectAllUsers() {
      if (this.allUsersSelected) {
        return (this.selectedUsers = [])
      }
      this.selectedUsers = [...this.users]
    },

    toggleSelectUser(toggledUser) {
      const isUserSelected = this.selectedUsers.find((user) => user.id === toggledUser.id)

      if (!isUserSelected) {
        return this.selectedUsers.push(toggledUser)
      }

      this.selectedUsers = this.selectedUsers.filter((user) => user.id !== toggledUser.id)
    },

    unselectAllUsers() {
      this.selectedUsers = []
    },

    toggleCreateUserModal() {
      this.createUserModalOpen = !this.createUserModalOpen
    },

    toggleDeleteUserModal() {
      this.deleteUserModalOpen = !this.deleteUserModalOpen
    },
    async deleteUsers(users) {
      const promises = users.map((user) => this.graphClient.users.deleteUser(user.id))

      try {
        await Promise.all(promises)
        this.showMessage({
          title: this.$gettextInterpolate(
            this.$ngettext(
              'User "%{user}" was deleted successfully',
              '%{userCount} users were deleted successfully',
              this.selectedUsers.length
            ),
            {
              userCount: this.selectedUsers.length,
              user: this.selectedUsers[0].onPremisesSamAccountName
            },
            true
          )
        })
        this.users = this.users.filter((user) => {
          return !users.find((deletedUser) => user.id === deletedUser.id)
        })
        this.selectedUsers = []
        this.toggleDeleteUserModal()
      } catch (error) {
        console.error(error)
        this.showMessage({
          title: this.$gettextInterpolate(
            this.$ngettext(
              'Failed to delete user "%{user}"',
              'Failed to delete %{userCount} users',
              this.selectedUsers.length
            ),
            {
              userCount: users.length,
              user: users.onPremisesSamAccountName
            },
            true
          ),
          status: 'danger'
        })
      }
    },

    async createUser(user) {
      try {
        const response = await this.graphClient.users.createUser(user)
        this.toggleCreateUserModal()
        this.showMessage({
          title: this.$gettext('User was created successfully')
        })
        this.users.push(response?.data)
      } catch (error) {
        console.error(error)
        this.showMessage({
          title: this.$gettext('Failed to create user'),
          status: 'danger'
        })
      }
    }
  }
}
</script>
