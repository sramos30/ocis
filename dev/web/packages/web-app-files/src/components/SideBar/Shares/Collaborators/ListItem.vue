<template>
  <div
    :data-testid="`collaborator-${isUser || isSpace ? 'user' : 'group'}-item-${
      share.collaborator.name
    }`"
    class="files-collaborators-collaborator oc-flex oc-flex-middle oc-py-xs oc-flex-between"
  >
    <div class="oc-width-2-3 oc-flex oc-flex-middle" style="gap: 10px">
      <avatar-image
        v-if="isUser || isSpace"
        :userid="share.collaborator.name"
        :user-name="share.collaborator.displayName"
        :width="48"
        class="files-collaborators-collaborator-indicator"
      />
      <oc-avatar-item
        v-else
        :width="48"
        icon-size="medium"
        :icon="shareTypeKey"
        :name="shareTypeKey"
        class="files-collaborators-collaborator-indicator"
      />
      <div class="oc-text-truncate">
        <p v-oc-tooltip="shareDisplayNameTooltip" class="oc-text-bold oc-text-truncate oc-m-rm">
          <span
            aria-hidden="true"
            class="files-collaborators-collaborator-name"
            v-text="shareDisplayName"
          />
          <span
            v-if="shareAdditionalInfo"
            aria-hidden="true"
            class="files-collaborators-collaborator-additional-info"
            v-text="shareAdditionalInfo"
          />
          <span class="oc-invisible-sr" v-text="screenreaderShareDisplayName" />
        </p>
        <p class="oc-m-rm oc-flex">
          <span
            aria-hidden="true"
            class="files-collaborators-collaborator-share-type"
            v-text="shareTypeText"
          />
          <span v-if="sharedParentRoute" class="oc-resource-indicators oc-text-truncate">
            <span class="oc-mx-s">Â·</span>
            <router-link
              v-oc-tooltip="$gettext('Navigate to parent folder')"
              class="parent-folder oc-text-truncate"
              :to="sharedParentRoute"
            >
              <span class="text" v-text="$gettext('via')" />
              <oc-icon name="folder-2" size="small" fill-type="line" class="oc-px-xs" />
              <span class="text oc-text-truncate" v-text="sharedParentDir" />
            </router-link>
          </span>
          <span class="oc-invisible-sr" v-text="screenreaderShareDetails" />
        </p>
        <p v-if="hasExpirationDate" class="oc-m-rm">
          <span
            v-oc-tooltip="expirationDate"
            aria-hidden="true"
            class="files-collaborators-collaborator-expiration"
            data-testid="recipient-info-expiration-date"
            tabindex="0"
            v-text="shareExpirationText"
          />
          <span class="oc-invisible-sr" v-text="screenreaderShareExpiration" />
        </p>
      </div>
    </div>
    <div
      v-if="canEditOrDelete"
      class="oc-width-1-3 oc-flex oc-flex-nowrap oc-flex-right oc-flex-middle"
    >
      <role-dropdown
        :resource="highlightedFile"
        :share-id="share.id"
        :existing-permissions="share.customPermissions"
        :existing-role="share.role"
        :allow-share-permission="hasResharing || isSpace"
        class="files-collaborators-collaborator-role"
        @optionChange="shareRoleChanged"
      />
      <edit-dropdown
        class="files-collaborators-collaborator-edit"
        data-testid="collaborator-edit"
        :expiration-date="share.expires ? share.expires : null"
        :share-category="shareCategory"
        @expirationDateChanged="shareExpirationChanged"
        @removeShare="removeShare"
      />
    </div>
    <div v-else-if="share.role">
      <span class="oc-mr-xs" v-text="share.role.label" />
    </div>
  </div>
</template>

<script>
import { mapGetters, mapActions, mapState } from 'vuex'
import Mixins from '../../../../mixins'
import { DateTime } from 'luxon'

import EditDropdown from './EditDropdown.vue'
import RoleDropdown from '../RoleDropdown.vue'
import { SharePermissions, ShareTypes } from '../../../../helpers/share'
import { clientService } from 'web-pkg/src/services'
import { useCapabilityFilesSharingResharing } from 'web-pkg/src/composables'

export default {
  name: 'ListItem',
  components: {
    EditDropdown,
    RoleDropdown
  },
  mixins: [Mixins],
  props: {
    share: {
      type: Object,
      required: true
    },
    modifiable: {
      type: Boolean,
      default: false
    },
    sharedParentRoute: {
      type: Object,
      default: null
    }
  },
  setup() {
    return {
      hasResharing: useCapabilityFilesSharingResharing()
    }
  },
  computed: {
    ...mapGetters('Files', ['highlightedFile']),
    ...mapGetters(['getToken', 'configuration']),
    ...mapState(['user']),

    shareType() {
      return ShareTypes.getByValue(this.share.shareType)
    },

    shareTypeKey() {
      return this.shareType.key
    },

    isGroup() {
      return this.shareType === ShareTypes.group
    },

    isUser() {
      return this.shareType === ShareTypes.user
    },

    isSpace() {
      return this.shareType === ShareTypes.space
    },

    shareTypeText() {
      return this.$gettext(this.shareType.label)
    },

    shareCategory() {
      return ShareTypes.isIndividual(this.shareType) ? 'user' : 'group'
    },

    shareDisplayName() {
      if (this.user.id === this.share.collaborator.name) {
        return this.$gettextInterpolate(this.$gettext('%{collaboratorName} (me)'), {
          collaboratorName: this.share.collaborator.displayName
        })
      }
      return this.share.collaborator.displayName
    },

    shareAdditionalInfo() {
      if (!this.share.collaborator.additionalInfo) {
        return
      }
      return ` (${this.share.collaborator.additionalInfo})`
    },

    shareDisplayNameTooltip() {
      return this.shareDisplayName + (this.shareAdditionalInfo || '')
    },

    screenreaderShareDisplayName() {
      const context = {
        displayName: this.share.collaborator.displayName,
        ...(this.share.collaborator.additionalInfo && {
          additionalInfo: this.share.collaborator.additionalInfo
        })
      }
      let translated = this.$gettext('Share receiver name: %{ displayName }')
      if (this.shareAdditionalInfo) {
        translated = this.$gettext('Share receiver name: %{ displayName } (%{ additionalInfo })')
      }
      return this.$gettextInterpolate(translated, context)
    },

    screenreaderShareDetails() {
      const translated = this.$gettext('Share type: %{ shareType }')
      return this.$gettextInterpolate(translated, { shareType: this.shareTypeText })
    },

    shareExpirationText() {
      const translated = this.$gettext('Expires %{ expiryDateRelative }')
      return this.$gettextInterpolate(translated, {
        expiryDateRelative: this.expirationDateRelative
      })
    },

    screenreaderShareExpiration() {
      const translated = this.$gettext('Share expires %{ expiryDateRelative } (%{ expiryDate })')
      return this.$gettextInterpolate(translated, {
        expiryDateRelative: this.expirationDateRelative,
        expiryDate: this.expirationDate
      })
    },

    canEditOrDelete() {
      return this.modifiable
    },

    hasExpirationDate() {
      return this.share.expires
    },

    expirationDate() {
      return DateTime.fromJSDate(this.share.expires)
        .endOf('day')
        .setLocale(this.$language.current)
        .toLocaleString(DateTime.DATETIME_FULL)
    },

    expirationDateRelative() {
      return DateTime.fromJSDate(this.share.expires)
        .endOf('day')
        .setLocale(this.$language.current)
        .toRelative()
    },

    graphClient() {
      return clientService.graphAuthenticated(this.configuration.server, this.getToken)
    },

    sharedParentDir() {
      return this.sharedParentRoute?.params?.item.split('/').pop()
    }
  },
  methods: {
    ...mapActions(['showMessage']),
    ...mapActions('Files', ['changeShare']),

    removeShare() {
      this.$emit('onDelete', this.share)
    },

    shareRoleChanged({ role, permissions }) {
      const expirationDate = this.share.expires
      try {
        this.saveShareChanges({ role, permissions, expirationDate })
      } catch (e) {
        console.error(e)
        this.showMessage({
          title: this.$gettext('Failed to apply new permissions'),
          status: 'danger'
        })
      }
    },

    shareExpirationChanged({ expirationDate }) {
      const role = this.share.role
      const permissions = this.share.customPermissions
      try {
        this.saveShareChanges({ role, permissions, expirationDate })
      } catch (e) {
        console.error(e)
        this.showMessage({
          title: this.$gettext('Failed to apply expiration date'),
          status: 'danger'
        })
      }
    },

    saveShareChanges({ role, permissions, expirationDate }) {
      const bitmask = role.hasCustomPermissions
        ? SharePermissions.permissionsToBitmask(permissions)
        : SharePermissions.permissionsToBitmask(
            role.permissions(this.hasResharing || this.shareType === ShareTypes.space)
          )
      this.changeShare({
        client: this.$client,
        graphClient: this.graphClient,
        share: this.share,
        permissions: bitmask,
        expirationDate: expirationDate || ''
      })
    }
  }
}
</script>

<style lang="scss" scoped>
.sharee-avatar {
  min-width: 48px;
}
</style>
