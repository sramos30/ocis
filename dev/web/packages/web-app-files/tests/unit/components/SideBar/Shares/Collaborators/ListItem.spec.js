import { createLocalVue, mount } from '@vue/test-utils'
import Vuex from 'vuex'
import ListItem from '../../../../../../src/components/SideBar/Shares/Collaborators/ListItem.vue'
import stubs from '../../../../../../../../tests/unit/stubs'
import GetTextPlugin from 'vue-gettext'
import { peopleRoleViewerFolder, ShareTypes } from '../../../../../../src/helpers/share'
import Users from '@/__fixtures__/users'

const localVue = createLocalVue()
localVue.use(Vuex)
localVue.use(GetTextPlugin, {
  translations: 'does-not-matter.json',
  silent: true
})

const selectors = {
  userAvatarImage: 'avatar-image-stub.files-collaborators-collaborator-indicator',
  notUserAvatar: 'oc-avatar-item-stub.files-collaborators-collaborator-indicator',
  collaboratorAdditionalInfo: '.files-collaborators-collaborator-additional-info',
  collaboratorName: '.files-collaborators-collaborator-name',
  shareType: '.files-collaborators-collaborator-share-type',
  collaboratorRole: '.files-collaborators-collaborator-role',
  collaboratorEdit: '.files-collaborators-collaborator-edit',
  shareInheritanceIndicators: '.oc-resource-indicators'
}

describe('Collaborator ListItem component', () => {
  describe('displays the correct image/icon according to the shareType', () => {
    describe('user share type', () => {
      it('should display a users avatar', () => {
        const wrapper = createWrapper({ shareType: ShareTypes.user.value })
        expect(wrapper.find(selectors.userAvatarImage).exists()).toBeTruthy()
        expect(wrapper.find(selectors.notUserAvatar).exists()).toBeFalsy()
      })
      it('sets user info on the avatar', () => {
        const wrapper = createWrapper()
        expect(wrapper.find(selectors.userAvatarImage).attributes('userid')).toEqual('brian')
        expect(wrapper.find(selectors.userAvatarImage).attributes('user-name')).toEqual(
          'Brian Murphy'
        )
      })
    })
    describe('non-user share types', () => {
      it.each(
        ShareTypes.all.filter(
          (shareType) => ![ShareTypes.user, ShareTypes.space].includes(shareType)
        )
      )('should display an oc-avatar-item for any non-user share types', (shareType) => {
        const wrapper = createWrapper({ shareType: shareType.value })
        expect(wrapper.find(selectors.userAvatarImage).exists()).toBeFalsy()
        expect(wrapper.find(selectors.notUserAvatar).exists()).toBeTruthy()
        expect(wrapper.find(selectors.notUserAvatar).attributes().name).toEqual(shareType.key)
      })
    })
  })
  describe('share info', () => {
    it('shows the collaborator display name', () => {
      const wrapper = createWrapper()
      expect(wrapper.find(selectors.collaboratorName).text()).toEqual('Brian Murphy')
    })
    describe('additionalInfo', () => {
      it('shows additional information about the collaborator if set', () => {
        const wrapper = createWrapper()
        expect(wrapper.find(selectors.collaboratorAdditionalInfo).text()).toEqual(
          '(brian@owncloud.com)'
        )
      })
      it('does not show additional information about the collaborator if not set', () => {
        const wrapper = createWrapper({
          collaborator: {
            displayName: 'Alice Hansen'
          }
        })
        expect(wrapper.find(selectors.collaboratorAdditionalInfo).exists()).toBeFalsy()
      })
    })
    it.each(ShareTypes.authenticated)('shows a label for the share type', (shareType) => {
      const wrapper = createWrapper({ shareType: shareType.value })
      expect(wrapper.find(selectors.shareType).text()).toBe(shareType.label)
    })
  })
  describe('modifiable property', () => {
    it('shows interactive elements when modifiable', () => {
      const wrapper = createWrapper({ modifiable: true })
      expect(wrapper.find(selectors.collaboratorRole).exists()).toBeTruthy()
      expect(wrapper.find(selectors.collaboratorEdit).exists()).toBeTruthy()
    })
    it('hides interactive elements when not modifiable', () => {
      const wrapper = createWrapper({ modifiable: false })
      expect(wrapper.find(selectors.collaboratorRole).exists()).toBeFalsy()
      expect(wrapper.find(selectors.collaboratorEdit).exists()).toBeFalsy()
    })
  })
  describe('share inheritance indicators', () => {
    it('show when sharedParentRoute is given', () => {
      const wrapper = createWrapper({ sharedParentRoute: { params: { item: '/folder' } } })
      expect(wrapper.find(selectors.shareInheritanceIndicators).exists()).toBeTruthy()
      expect(wrapper).toMatchSnapshot()
    })
    it('do not show when sharedParentRoute is not given', () => {
      const wrapper = createWrapper()
      expect(wrapper.find(selectors.shareInheritanceIndicators).exists()).toBeFalsy()
    })
  })
})

function createWrapper({
  shareType = ShareTypes.user.value,
  collaborator = {
    name: 'brian',
    displayName: 'Brian Murphy',
    additionalInfo: 'brian@owncloud.com'
  },
  role = peopleRoleViewerFolder,
  modifiable = true,
  sharedParentRoute = null
} = {}) {
  return mount(ListItem, {
    store: new Vuex.Store({
      state: {
        user: Users.alice
      },
      modules: {
        Files: {
          namespaced: true,
          getters: {
            highlightedFile: function () {
              return { type: 'folder', isFolder: true }
            }
          }
        }
      }
    }),
    propsData: {
      share: {
        collaborator,
        shareType,
        role
      },
      modifiable,
      sharedParentRoute
    },
    localVue,
    stubs: {
      ...stubs,
      'oc-table-simple': true,
      'oc-tr': true,
      'oc-td': true,
      'oc-tag': true,
      'oc-pagination': true,
      'oc-avatar-item': true,
      'role-dropdown': true,
      'edit-dropdown': true,
      translate: false
    },
    directives: {
      'oc-tooltip': jest.fn()
    }
  })
}
