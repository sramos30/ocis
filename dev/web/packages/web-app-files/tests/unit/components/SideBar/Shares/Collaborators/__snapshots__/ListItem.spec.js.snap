// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Collaborator ListItem component share inheritance indicators show when sharedParentRoute is given 1`] = `
<div data-testid="collaborator-user-item-brian" class="files-collaborators-collaborator oc-flex oc-flex-middle oc-py-xs oc-flex-between">
  <div class="oc-width-2-3 oc-flex oc-flex-middle" style="gap: 10px;">
    <avatar-image-stub userid="brian" user-name="Brian Murphy" width="48" class="files-collaborators-collaborator-indicator"></avatar-image-stub>
    <div class="oc-text-truncate">
      <p class="oc-text-bold oc-text-truncate oc-m-rm"><span aria-hidden="true" class="files-collaborators-collaborator-name">Brian Murphy</span> <span aria-hidden="true" class="files-collaborators-collaborator-additional-info"> (brian@owncloud.com)</span> <span class="oc-invisible-sr">Share receiver name: Brian Murphy (brian@owncloud.com)</span></p>
      <p class="oc-m-rm oc-flex"><span aria-hidden="true" class="files-collaborators-collaborator-share-type">User</span> <span class="oc-resource-indicators oc-text-truncate"><span class="oc-mx-s">Â·</span>
        <router-link-stub to="[object Object]" class="parent-folder oc-text-truncate"><span class="text">via</span>
          <oc-icon-stub name="folder-2" size="small" fill-type="line" class="oc-px-xs"></oc-icon-stub> <span class="text oc-text-truncate">folder</span>
        </router-link-stub>
        </span> <span class="oc-invisible-sr">Share type: User</span>
      </p>
      <!---->
    </div>
  </div>
  <div class="oc-width-1-3 oc-flex oc-flex-nowrap oc-flex-right oc-flex-middle">
    <role-dropdown-stub resource="[object Object]" existingrole="[object Object]" existingpermissions="" class="files-collaborators-collaborator-role"></role-dropdown-stub>
    <edit-dropdown-stub sharecategory="user" data-testid="collaborator-edit" class="files-collaborators-collaborator-edit"></edit-dropdown-stub>
  </div>
</div>
`;
